### PR TITLE
POC Dict merging/concatenating

### PR DESCRIPTION
**Pull request by Axik**
https://github.com/Deepwalker/trafaret/issues/30

POC of working merging/concatenating Dict trafarets. 
Review please and let's make conclusion on this. I'm open for any suggestion about naming conventions and message text etc. Also we can discuss API of this. 

@jettify, @Deepwalker , @asvetlov

### CODE REVIEW

File: trafaret/__init__.py [hunk 1]
```diff
@@ -976,6 +976,11 @@ def __init__(self, keys={}, **trafarets):
             key_.set_trafaret(self._trafaret(trafaret))
             self.keys.append(key_)
 
+    def _to_names(self):
+        for key in self.keys:
+            if key.to_name:
+                yield key.to_name
```
Comment (Deepwalker) [inline trafaret/__init__.py]: class Key is only one of possible keys – you cant use its internal var.

Comment (jettify) [inline trafaret/__init__.py]: Could you elaborate on this?

**Comment by jettify** [comment]
Any progress on this? This is very useful feature for me...

**Comment by Axik** [comment]
I would propose that consider to compare only key names and no `to_name`.

File: trafaret/__init__.py [hunk 2]
```diff
@@ -1046,6 +1052,18 @@ def __repr__(self):
         r += ")>"
         return r
 
+    def __and__(self, other):
+
+        if set(self.keys_names()) & set(other.keys_names()):
+            raise DataError('Dicts that merging should '
```
Comment (tymofij) [inline trafaret/__init__.py]: IMO better English would be "Merged dicts should have no interlapping keys"

**Comment by Axik** [comment]
@nimnull review it please

File: trafaret/__init__.py (hunk 1 above)
Comment (nimnull) [inline trafaret/__init__.py]: @Deepwalker according to https://github.com/Axik/trafaret/blob/master/trafaret/__init__.py#L974:

``` python
        for key, trafaret in itertools.chain(trafarets.items(), keys.items()):
            key_ = key if isinstance(key, Key) else Key(key)
            key_.set_trafaret(self._trafaret(trafaret))
            self.keys.append(key_)
```

All key definitions passed to the new `Dict` instance processed/converted to instance of `Key`. So `Key` instance is only possible key for `Dict`.

File: trafaret/__init__.py [hunk 3]
```diff
@@ -1046,6 +1052,18 @@ def __repr__(self):
         r += ")>"
         return r
 
+    def __and__(self, other):
+
```
Comment (asvetlov) [inline trafaret/__init__.py]: Please add `isinstance` check for `other` param.

File: trafaret/__init__.py [hunk 4]
```diff
@@ -1046,6 +1046,21 @@ def __repr__(self):
         r += ")>"
         return r
 
+    def __and__(self, other):
+        if not isinstance(other, Dict):
+            raise DataError('You must merge Dict only with Dict')
```
Comment (Deepwalker) [inline trafaret/__init__.py:1051]: TypeError

File: trafaret/__init__.py [hunk 5]
```diff
@@ -1046,6 +1046,21 @@ def __repr__(self):
...
+        if set(self.keys_names()) & set(other.keys_names()):
+            raise DataError('Merged dicts should have '
```
Comment (Deepwalker) [inline trafaret/__init__.py:1053]: May be ValueError, not sure

File: trafaret/__init__.py [hunk 6]
```diff
@@ -1046,6 +1046,21 @@ def __repr__(self):
...
+                            'no interlapping keys')
+        if set(
+            key.get_name() for key in self.keys) & set(
+                key.get_name() for key in other.keys):
+            raise DataError('Merged dicts should have '
```
Comment (Deepwalker) [inline trafaret/__init__.py:1058]: Some here

**Comment by rrader** [comment]
May I ask why "&" instead of "+"? For me merging would be more intuitive by adding two trafarets instead of using bitwise-and operator.

**Comment by asvetlov** [comment]
Good point

**Pull request merged**